### ISSUE TITLE
Conflicts with Magento's Native Elasticsuite Module?

### ISSUE THREAD

**Issue by dambrogia** [Smile-SA/elasticsuite#2234]
I'm running into indexing throughput issues when making updates to categories after updating to Magento 2.4 and smile/elasticsuite version 2.10.3.

I can replicate this by running the following in a development environment (using the `n98-magerun` tool for convenience/simplicity):

    n98-magerun db:query -- "INSERT INTO magento.catalog_category_product_cl (entity_id) VALUES (178)"; 

You can replace 178 with whatever category entity id you want, you should have 1 item in the backlog of your `catalog_category_product` mview at this point (assuming you started with none in there).

Then run the mview cron job:

    n98-magerun sys:cron:run indexer_update_all_views

I notice the usual logic taking place, I can follow the data go into the replica + index tables.  The process is mostly efficient, a lot of data is written but it is done so in bulk with decent (mysql) queries and it all finishes within a few seconds.

What caught me off guard was the 5800 products in the category (ID: 178) being indexed from the file [`magento/module-catalog/Model/Indexer/Category/Product/AbstractAction.php`](https://github.com/magento/magento2/blob/445b0f1a3d6a91b5da32d311077c2112ef0b1503/app/code/Magento/Elasticsearch/Model/Indexer/Fulltext/Plugin/Category/Product/Action/Rows.php#L77-L89). The plugin ID/name is `catalogsearchFulltextProductAssignment`.

Within the the code from the link above, the `Fulltext::INDEXER_ID` is `catalogsearch_fulltext` which this library takes control over.

___

So my questions are:

- Is this magento-native indexing action necessary when using the smile/elasticsuite module?
- Is it necessary for all categories? or just non-virtual categories?
- Is this action explicitly for placing products in categories? 
    - Would it then be possible increase performance with an "around" interceptor that does the following?
        - Store product ID's that belong to the in-scope category in a temp table before any updates occur (pre-state)
        - Call the continue function to run default/existing logic
        - Index any product ID's that exist in the new state of the category and not the temp table OR exist in the temp-table but not the current state of the category.

I'm not asking for a change to this repository (at the moment), just insight on how to best follow design patterns with your existing logic and how it integrates into Magento's native logic. If I find something useful, I would be glad to share my findings with the community.

Thank you for the help.

**Comment by dambrogia** [comment]
I wanted to give back some updates from what I've found after debugging this issue for a few days.

> - Is this magento-native indexing action necessary when using the smile/elasticsuite module?

Yes and No.

There is another native plugin (`catalogsearchFulltextCategoryAssignment`) that will handle product/category assignments (the plugin originally mentioned handles category/products assignments). This means that as you add products to your non-virtual category, their native interceptor/plugin will take the newly added/removed products and correctly and efficiently index their new category assignment into the catalogsearch_fulltext index. This plugin does _nothing_ for virtual categories, since when the products for categories are read natively, the virtual-rule is not loaded/rendered/read causing only non-virtual category products to show up.

So this technically is required for virtual categories, but from what I am vaguely starting to understand -- it works extremely inefficiently by just reindexing the entire root category which is the only way your products get reindexed into the fulltext index. Since the changelog has no concept of scope, etc and the virtual rules for categories require the search context (`Smile\ElasticsuiteCore\Api\Search\ContextInterface`) the only way to implement some sort of context as you move through these categories in the chanagelog would be to set the context per-category (please correct me if I'm wrong).

So TLDR;
 - It is not required for non-virtual categories
 - it is required for virtual category updates
 - It is wildly inefficient and inaccurate for the products _needing_ to be updated for virtual categories

> - Is it necessary for all categories? or just non-virtual categories?

Based on the TLDR; above ^, only required for virtual since non-virtual can be handled by `catalogsearchFulltextCategoryAssignment`, but [`catalogsearchFulltextProductAssignment`](https://github.com/magento/magento2/blob/445b0f1a3d6a91b5da32d311077c2112ef0b1503/app/code/Magento/Elasticsearch/etc/di.xml#L19) is wildly inefficient.

> - Is this action explicitly for placing products in categories?

Yes, this is intended for category/product relationships

> - Would it then be possible increase performance with an "around" interceptor that does the following?
>     - Store product ID's that belong to the in-scope category in a temp table before any updates occur (pre-state)
>     - Call the continue function to run default/existing logic
>     - Index any product ID's that exist in the new state of the category and not the temp table OR exist in the temp-table but not the current state of the category.

Adding _just_ this simple change will not work because the products being loaded here are for non-virtual categories. Virtual Rules are not read from to generate their products. And the entire goal here is to reindex _necessary_ products into the fulltext index.

We are now looking at solving this in one of two ways, both generally require reading the virtual categories and reindexing their products in the changelog.

1. Use the `catalog_category_save_after` event in the admin. (since there is a provided scope/context)
2. Continue using the changelog, ignore the first 2 levels of the Category Path (usually 1/2 for the default store) to prevent unnecessary products and update the scope per category processed.

@rbayet I'm pinging you, since you were assigned the issue. But I am having trouble rendering products for a given virtual category. I have searched the docs and the codebase pretty thoroughly.

The closest snippet of code I have produced is:

```

/** @var Smile\ElasticsuiteCore\Api\Search\ContextInterface */
$this->elasticSearchContext->setCurrentCategory($category);
/** @var Smile\ElasticsuiteCatalog\Model\Category\Filter\Provider */
$query = $queryProvider->getQueryFilter($category);
/** @var Smile\ElasticsuiteCatalog\Model\ResourceModel\Product\Fulltext\CollectionFactory */
$products = $fulltextCollection->create()->addQueryFilter($query)->getItems();


```

However I get an error about the index not existing. _But_ I have reindexed the catalogsearch_fulltext index to create the index, I can preview categories in the admin and load category pages so I think my code is wrong rather than the index not existing.

Any code samples or references to code / documentation would be greatly appreciated.

**Comment by romainruaud** [comment]
There should be no need to index something when saving/updating a virtual category : the rule belonging to the virtual category is computed real-time in frontend when accessing the category. And no association between product and the so-called virtual category is stored anywhere.

=> I don't understand why reindexing the products should be needed in this case ? 

However, I have a feeling that something is broken with latest Magento versions : #2162 (and I tend to think our associated fix #2171 is not enough actually). 

I can reproduce this, updating a virtual category rule in backend does not cause the previously computed (and cached) rule to expire, that might lead to some inconsistencies : the preview in the back-office does not use the cache (obviously), but the frontend result does.

That being said, you snippet to fetch products assigned to a virtual category seems to be correct for me, if the index is not existing, it might be due to a missing `setStoreId()` to the collection, where the code might try to query an index associated to  the store "0" which will never exists : we only create an index per "real" store-view, not for the admin (0).

Regards

**Comment by dambrogia** [comment]
Sorry for the late response,but I do want to confirm that this was purely a cache related issue and we previously resolved it by clearing the cache for the virtual rule when saving a virtual category (just as you had mentioned) and ensuring there was no indexing of products during virtual category saves (because also as you mentioned, we realized this happens in real time).

Thank you for the insight, I can close this now (selfishly, my issue is solved) or if there is any context/fixes you would like help with I can dig up our commits for this issue.
